framework/src/materials/GenericFunctorMaterial.C

```diff
@@ -0,0 +1,84 @@
+registerMooseObjectRenamed("MooseApp",
+                           GenericConstantFunctorMaterial,
+                           "06/30/2022 24:00",
+                           GenericFunctorMaterial);
+registerMooseObjectRenamed("MooseApp",
+                           ADGenericConstantFunctorMaterial,
+                           "06/30/2022 24:00",
+                           ADGenericFunctorMaterial);
+registerMooseObjectRenamed("MooseApp",
+                           GenericFunctionFunctorMaterial,
+                           "06/30/2022 24:00",
+                           GenericFunctorMaterial);
+registerMooseObjectRenamed("MooseApp",
+                           ADGenericFunctionFunctorMaterial,
+                           "06/30/2022 24:00",
+                           ADGenericFunctorMaterial);
+
+template <bool is_ad>
+InputParameters
+GenericFunctorMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += SetupInterface::validParams();
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
+  params.addClassDescription(
+      "FunctorMaterial object for declaring properties that are populated by evaluation of a "
+      "Functor (a constant, variable, function or functor material property) objects.");
+  params.addParam<std::vector<std::string>>("prop_names",
+                                            "The names of the properties this material will have");
+  params.addParam<std::vector<MooseFunctorName>>("prop_values",
+                                                 "The corresponding names of the "
+                                                 "functors that are going to provide "
+                                                 "the values for the variables");
+  return params;
+}
+
+template <bool is_ad>
+GenericFunctorMaterialTempl<is_ad>::GenericFunctorMaterialTempl(const InputParameters & parameters)
+  : FunctorMaterial(parameters),
+    _prop_names(getParam<std::vector<std::string>>("prop_names")),
+    _prop_values(getParam<std::vector<MooseFunctorName>>("prop_values"))
+{
+  unsigned int num_names = _prop_names.size();
+  unsigned int num_values = _prop_values.size();
+
+  if (num_names != num_values)
+    mooseError("Number of prop_names must match the number of prop_values for a "
+               "GenericFunctorMaterial!");
+
+  _num_props = num_names;
+  _functors.resize(num_names);
+
+  for (const auto i : make_range(_num_props))
+    _functors[i] = &getFunctor<GenericReal<is_ad>>(_prop_values[i]);
+
+  const std::set<ExecFlagType> clearance_schedule(_execute_enum.begin(), _execute_enum.end());
+  for (const auto i : make_range(_num_props))
+  {
+    auto & prop = declareFunctorProperty<GenericReal<is_ad>>(_prop_names[i]);
+    prop.setFunctor(
+        _mesh, blockIDs(), [this, i](const auto & r, const auto & t) -> GenericReal<is_ad> {
+          return (*_functors[i])(r, t);
+        });
```

Comment:
The power of functors! Look at this arbitrary evaluation chaining!!

Reply:
But now I am wondering ... if we are just forwarding evaluation to another functor, should *this* functor even exist? What new thing is it doing that the forwarded-to functor didn't do?

Reply:
it changes the name lol

Reply:
and in the future, it may or may not be able to change the AD type

Reply:
oh and it can change the caching behavior!

Reply:
I m just trying to get my PR in tbh. If you want a deletion i ll do it

Reply:
Well you're doing the kind act of consolidating ... so I'll let this go for now. At a minimum you did not add any new unnecessary objects

Reply:
actually it's cool to do something like that:
```
  [firebrick_properties]
    type = ADGenericFunctorMaterial
    prop_names = 'rho_s         cp_s        k_s'
    prop_values = '${solid_rho} ${solid_cp} 0.26'
    block = '10'
  []
 ```
some kernels and materials in Pgh just like to retrieve the functors directly, not in the parameters but through the PGH namespace